From 7952a3f45464ac5e22c5144bfff75910f3466bac Mon Sep 17 00:00:00 2001
From: Rodney Dawes <dobey.pwns@gmail.com>
Date: Thu, 27 Sep 2018 18:00:39 -0400
Subject: [PATCH 1/3] Remove dependency on ual.

The dependency on ubuntu-app-launch creates a circular dependency, as
ubuntu-app-launch depends on libertine, which depends on system-settings.
Instead, we'll just build the URL and use QDesktopServices to open it.
---
 CMakeLists.txt                               |  1 +
 debian/control                               |  1 -
 plugins/system-update/CMakeLists.txt         | 11 ++++++-----
 plugins/system-update/click/manager_impl.cpp | 15 ++++-----------
 tests/plugins/system-update/CMakeLists.txt   |  5 +----
 5 files changed, 12 insertions(+), 21 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index daa2b58b..c45d28de 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,6 +32,7 @@ endif()
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x -fno-permissive -pedantic -Wall -Wextra")
 find_package(Qt5Concurrent REQUIRED)
 find_package(Qt5DBus REQUIRED)
+find_package(Qt5Gui REQUIRED)
 find_package(Qt5Qml REQUIRED)
 find_package(Qt5Quick REQUIRED)
 find_package(Qt5Sql REQUIRED)
diff --git a/debian/control b/debian/control
index 99e0364c..5a4cfde3 100644
--- a/debian/control
+++ b/debian/control
@@ -21,7 +21,6 @@ Build-Depends: dbus-test-runner,
                libpolkit-agent-1-dev,
                libqmenumodel-dev,
                libtrust-store-dev,
-               libubuntu-app-launch2-dev,
                libudev-dev,
                libunity-api-dev,
                libupower-glib-dev,
diff --git a/plugins/system-update/CMakeLists.txt b/plugins/system-update/CMakeLists.txt
index 5cee727e..57488cc5 100644
--- a/plugins/system-update/CMakeLists.txt
+++ b/plugins/system-update/CMakeLists.txt
@@ -20,9 +20,6 @@ include_directories(
     /usr/include/apt-pkg/
 )
 
-pkg_check_modules(UAL REQUIRED ubuntu-app-launch-2)
-add_definitions(${UAL_CFLAGS} ${UAL_CFLAGS_OTHER})
-
 add_library(UpdatePlugin SHARED
     click/apiclient.h
     click/manager.h
@@ -46,10 +43,14 @@ add_library(UpdatePlugin SHARED
     ../../src/i18n.cpp
 )
 target_link_libraries(UpdatePlugin
-    Qt5::Quick Qt5::Core Qt5::Network Qt5::Sql
+    Qt5::Core
+    Qt5::Gui
+    Qt5::Network
+    Qt5::Quick
+    Qt5::Sql
+
     apt-pkg
     uss-systemimage
-    ${UAL_LDFLAGS}
 )
 
 add_library(UbuntuUpdatePanel MODULE
diff --git a/plugins/system-update/click/manager_impl.cpp b/plugins/system-update/click/manager_impl.cpp
index 25769e6d..b23c641a 100644
--- a/plugins/system-update/click/manager_impl.cpp
+++ b/plugins/system-update/click/manager_impl.cpp
@@ -23,9 +23,8 @@
 #include "network/accessmanager_impl.h"
 #include "../../src/i18n.h"
 
-#include <ubuntu-app-launch.h>
-
 #include <QDateTime>
+#include <QDesktopServices>
 #include <QFinalState>
 #include <QState>
 #include <QJsonDocument>
@@ -212,15 +211,9 @@ void ManagerImpl::cancel()
 
 bool ManagerImpl::launch(const QString &identifier)
 {
-    bool success = false;
-    gchar *appId = ubuntu_app_launch_triplet_to_app_id(
-        identifier.toLatin1().data(), nullptr, nullptr
-    );
-    if (appId) {
-        success = ubuntu_app_launch_start_application(appId, nullptr);
-    }
-    g_free(appId);
-    return success;
+    auto url = QStringLiteral(
+        "appid://%1/first-listed-app/current-user-version").arg(identifier);
+    return QDesktopServices::openUrl(url);
 }
 
 void ManagerImpl::handleManifest(const QJsonArray &manifest)
diff --git a/tests/plugins/system-update/CMakeLists.txt b/tests/plugins/system-update/CMakeLists.txt
index 6b1ab809..816c4fb3 100644
--- a/tests/plugins/system-update/CMakeLists.txt
+++ b/tests/plugins/system-update/CMakeLists.txt
@@ -7,15 +7,12 @@ include_directories(
     ${QTDBUSTEST_INCLUDE_DIRS}
 )
 
-pkg_check_modules(UAL REQUIRED ubuntu-app-launch-2)
-add_definitions(${UAL_CFLAGS} ${UAL_CFLAGS_OTHER})
-
 find_package(Qt5Test REQUIRED)
 find_package(Qt5Sql REQUIRED)
 
 set(CMAKE_AUTOMOC ON)
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
-set(PLUGIN_LIBS UpdatePlugin Qt5::Test Qt5::Sql ${UAL_LDFLAGS})
+set(PLUGIN_LIBS UpdatePlugin Qt5::Test Qt5::Sql)
 
 file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/mockclickserver.py DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
 
-- 
2.26.2

